lib/utils/compose_ts.ts

```diff
@@ -131,6 +130,372 @@ async function resolveProject(
 	return [composeFileName, composeFileContents];
 }
 
+interface BuildTaskPlus extends MultiBuild.BuildTask {
+	logBuffer?: string[];
+}
+
+interface Renderer {
+	start: () => void;
+	end: (buildSummaryByService?: Dictionary<string>) => void;
+	streams: Dictionary<NodeJS.ReadWriteStream>;
+}
+
+export async function buildProject(opts: {
+	docker: Dockerode;
+	logger: Logger;
+	projectPath: string;
+	projectName: string;
+	composition: Composition;
+	arch: string;
+	deviceType: string;
+	emulated: boolean;
+	buildOpts: import('./docker').BuildOpts;
+	inlineLogs?: boolean;
+	convertEol: boolean;
+	dockerfilePath?: string;
+	nogitignore: boolean;
+	multiDockerignore: boolean;
+}): Promise<BuiltImage[]> {
```

Comment:
`buildProject` was very large single function written in a `.then` promise chain. Such large functions are harder to test (if we wanted to...) and harder to make sense of, in my opinion. I've split it in some 7 thematic smaller functions, mostly mapping to `.then()` blocks of the old promise chain. Even then, what was left in `buildProject` itself is still of a significant size!